addons/binding/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/protocol/RemoteControllerWebSocket.java

```diff
@@ -0,0 +1,678 @@
+        jsonParams.put("Option", "false");
+        jsonParams.put("TypeOfRemote", "SendRemoteKey");
+
+        JSONObject json = new JSONObject();
+        json.put("method", "ms.remote.control");
+        json.put("params", jsonParams);
+
+        webSocketRemote.sendCommand(json.toJSONString());
+    }
+
+    public void sendSourceApp(String app) {
+        String appName = app;
+        App appVal = apps.get(app);
+        boolean deepLink = false;
+        if (appVal != null) {
+            appName = appVal.appId;
+            deepLink = appVal.type == 2;
+        }
+
+        JSONObject jsonData = new JSONObject();
+        jsonData.put("appId", appName);
+        jsonData.put("action_type", deepLink ? "DEEP_LINK" : "NATIVE_LAUNCH");
+
+        JSONObject jsonParams = new JSONObject();
+        jsonParams.put("event", "ed.apps.launch");
+        jsonParams.put("to", "host");
+        jsonParams.put("data", jsonData);
+
+        JSONObject json = new JSONObject();
+        json.put("method", "ms.channel.emit");
+        json.put("params", jsonParams);
+
+        webSocketRemote.sendCommand(json.toJSONString());
+    }
+
+    public void sendUrl(String url) {
+
+        JSONObject jsonData = new JSONObject();
+        jsonData.put("appId", "org.tizen.browser");
+        jsonData.put("action_type", "NATIVE_LAUNCH");
+        jsonData.put("metaTag", url);
+
+        JSONObject jsonParams = new JSONObject();
+        jsonParams.put("event", "ed.apps.launch");
+        jsonParams.put("to", "host");
+        jsonParams.put("data", jsonData);
+
+        JSONObject json = new JSONObject();
+        json.put("method", "ms.channel.emit");
+        json.put("params", jsonParams);
+
+        webSocketRemote.sendCommand(json.toJSONString());
+    }
+
+    private void getApps() {
+        getInfo("ed.installedApp.get");
+    }
+
+    private void getInfo(String str) {
+        JSONObject jsonParams = new JSONObject();
+        jsonParams.put("event", str);
+        jsonParams.put("to", "host");
+
+        JSONObject json = new JSONObject();
+        json.put("method", "ms.channel.emit");
+        json.put("params", jsonParams);
+
+        webSocketRemote.sendCommand(json.toJSONString());
+    }
+
+    public List<String> getAppList() {
+        ArrayList<String> appList = new ArrayList<>();
+        for (App app : apps.values()) {
+            appList.add(app.name);
+        }
+        return appList;
+    }
+
+    @Override
+    public void lifeCycleStarted(LifeCycle arg0) {
+        connectWebSockets();
+    }
+
+    @Override
+    public void lifeCycleFailure(LifeCycle arg0, Throwable throwable) {
+        logger.warn("Problem creating websocket client", throwable);
+    }
+
+    @Override
+    public void lifeCycleStarting(LifeCycle arg0) {
+    }
+
+    @Override
+    public void lifeCycleStopped(LifeCycle arg0) {
+    }
+
+    @Override
+    public void lifeCycleStopping(LifeCycle arg0) {
+    }
+
+}

```

Comment:
This class is very long, maybe it would make sense to split of the inner classes.

Reply:
Done